### PR TITLE
add kubelet gpu affinity

### PR DESCRIPTION
**Pull request by guang7niu**
**What type of PR is this?**
> Uncomment only one, leave it on its own line:
>
> /kind api-change
> /kind bug
> /kind cleanup
> /kind design
> /kind documentation
> /kind failing-test
> /kind feature
> /kind flake

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note

```

### CODE REVIEW

File: pkg/kubelet/cm/devicemanager/rank_test.go [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+	available.Insert("0", "4", "5", "6", "7")
+	ret = mgr.calcAllocated("nvidia.com/gpu", 6, sets.String{}, available)
+	sort.Sort(sort.StringSlice(ret))
+	v = strings.Join(ret, ",")
+	klog.V(2).Infoln("========== result", v)
+	if v != "" {
+		t.FailNow()
+	}
+
+	klog.V(2).Infoln("================= All tests done ==================")
```
Comment (killwing) [inline pkg/kubelet/cm/devicemanager/rank_test.go]: available: [0 2 4] needed: 2
result: [0, 4] ?

File: pkg/kubelet/cm/devicemanager/rank.go [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+
+const (
+	MAXCOST = 10000
+)
+
+var costs [][]int = costs_()
+
+func costs_() [][]int { // communication cost between gpus
+	return [][]int{
+		{0, 1, 10, 10, 100, 100, 100, 100},
```
Comment (killwing) [inline pkg/kubelet/cm/devicemanager/rank.go]: gpu topo info can be retrieved by:
github.com/NVIDIA/gpu-monitoring-tools/bindings/go/nvml

File: pkg/kubelet/cm/devicemanager/rank.go [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+import (
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/klog"
+)
+
+const (
+	MAXCOST = 10000
+)
+
+var costs [][]int = costs_()
```
Comment (killwing) [inline pkg/kubelet/cm/devicemanager/rank.go]: values can be assigned directly

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 4]
```diff
@@ -76,6 +76,9 @@ type ManagerImpl struct {
 	// e.g. a new device is advertised, two old devices are deleted and a running device fails.
 	callback monitorCallback
 
+	// SM
```
Comment (killwing) [inline pkg/kubelet/cm/devicemanager/manager.go]: meaningless comment

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 5]
```diff
@@ -142,7 +146,10 @@ func (m *ManagerImpl) genericDeviceUpdateCallback(resourceName string, devices [
 	m.mutex.Lock()
 	m.healthyDevices[resourceName] = sets.NewString()
 	m.unhealthyDevices[resourceName] = sets.NewString()
-	for _, dev := range devices {
+	for idx, dev := range devices { // SM
+		if resourceName == "nvidia.com/gpu" { // SM
```
Comment (killwing) [inline pkg/kubelet/cm/devicemanager/manager.go]: it's better to define a global constant for "nvidia.com/gpu"

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 6]
```diff
@@ -615,7 +622,9 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
 	if int(available.Len()) < needed {
 		return nil, fmt.Errorf("requested number of devices unavailable for %s. Requested: %d, Available: %d", resource, needed, available.Len())
 	}
-	allocated := available.UnsortedList()[:needed]
+	// allocated := available.UnsortedList()[:needed] // SM
```
Comment (killwing) [inline pkg/kubelet/cm/devicemanager/manager.go]: useless comment

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 7]
```diff
@@ -615,7 +622,9 @@ func (m *ManagerImpl) devicesToAllocate(podUID, contName, resource string, requi
...
+	allocated := m.calcAllocated(resource, needed, devicesInUse, available)
```
Comment (killwing) [inline pkg/kubelet/cm/devicemanager/manager.go]: not consider unhealthyDevices ?

File: pkg/kubelet/cm/devicemanager/rank_test.go [hunk 8]
```diff
@@ -0,0 +1,343 @@
...
+	klog.InitFlags(flag.CommandLine)
+	flag.StringVar(&logLevel, "logLevel", "3", "test")
+	flag.Lookup("v").Value.Set(logLevel)
+
+	mgr := &ManagerImpl{gpuBitmap: make(map[string]int, 0)}
+	for i := 0; i < 8; i++ {
+		mgr.gpuBitmap[strconv.Itoa(i)] = i
+	}
+
+	klog.V(2).Infoln("================= Begin tests ==================")
```
Comment (killwing) [inline pkg/kubelet/cm/devicemanager/rank_test.go]: useless comment

File: pkg/kubelet/cm/devicemanager/rank_test.go [hunk 9]
```diff
@@ -0,0 +1,343 @@
...
+	// 0-7, need 1
+	available = make(sets.String, 0)
+	available.Insert("0", "1", "2", "3", "4", "5", "6", "7")
+	ret = mgr.calcAllocated("nvidia.com/gpu", 1, sets.String{}, available)
+	sort.Sort(sort.StringSlice(ret))
+	v = strings.Join(ret, ",")
+	klog.V(2).Infoln("========== result", v)
```
Comment (killwing) [inline pkg/kubelet/cm/devicemanager/rank_test.go]: use t.Log in test

File: pkg/kubelet/cm/devicemanager/manager.go [hunk 10]
```diff
@@ -76,6 +76,9 @@ type ManagerImpl struct {
 	// e.g. a new device is advertised, two old devices are deleted and a running device fails.
 	callback monitorCallback
 
+	// SM
+	gpuBitmap map[string]int
```
Comment (houz42) [inline pkg/kubelet/cm/devicemanager/manager.go]: map[string]bool

File: pkg/kubelet/cm/devicemanager/rank.go [hunk 11]
```diff
@@ -0,0 +1,244 @@
...
+	lcost, lidx, ln := doRanking(n.left, num)
+	rcost, ridx, rn := doRanking(n.right, num)
+	klog.V(3).Infoln("SMAFFINITY ranking left", num, lcost, lidx, ln)
+	klog.V(3).Infoln("SMAFFINITY ranking right", num, rcost, ridx, rn)
+
+	if ln == nil && rn == nil { // neither satisfied
+		return MAXCOST, nil, nil
+	} else if ln != nil && rn != nil && ln != rn { // both satisfied
+		if lcost == rcost { // check parent recursively
+			for true {
```
Comment (houz42) [inline pkg/kubelet/cm/devicemanager/rank.go]: true 可以省略

File: pkg/kubelet/cm/devicemanager/rank.go [hunk 12]
```diff
@@ -0,0 +1,244 @@
...
+				ln, rn = ln.parent, rn.parent
+				if ln != nil && rn != nil && ln != rn {
```
Comment (houz42) [inline pkg/kubelet/cm/devicemanager/rank.go]:
```go
if ln == rn {
    return ln
}
```
?

File: pkg/kubelet/cm/devicemanager/rank.go [hunk 13]
```diff
@@ -0,0 +1,244 @@
...
+		max = max / 2
+	}
+	return max
+}
+
+type node struct {
+	parent *node
+	left   *node
+	right  *node
+	bm     []bool // globally shared, updated while ranking
```
Comment (houz42) [inline pkg/kubelet/cm/devicemanager/rank.go]: bm means idle devices?

**Comment by killwing** [comment]
transfer to https://github.com/supremind/kubernetes
also check https://github.com/supremind/kubernetes/pull/1

**Pull request closed without merging**